src/framework/components/esmscript/component.js

```diff
@@ -0,0 +1,525 @@
+     *
+     * @param {string} moduleSpecifier - The module specifier to search for
+     * @returns {boolean} If script is attached to an entity.
+     * @example
+     * if (entity.module.has('path/to/module.mjs')) {
+     *     // entity has script
+     * }
+     */
+    has(moduleSpecifier) {
+        return this.modules.has(moduleSpecifier);
+    }
+
+    /**
+     * Get a script instance (if attached).
+     *
+     * @param {string} moduleSpecifier - The
+     * name or type of {@link ScriptType}.
+     * @returns {import('../../script/script-type.js').ScriptType|null} If script is attached, the
+     * instance is returned. Otherwise null is returned.
+     * @example
+     * const controller = entity.module.get('module');
+     */
+    get(moduleSpecifier) {
+        return this.modules.get(moduleSpecifier);
+    }
+
+    /**
+     * Registers an in-memory Script ES module with the the component system, allowing it to interact with the engine update mechanic.
+     * If the module already exists, ie. a module with the same specifier exists in this component, then it will
+     * replace this *if* the module has a swap method.
+     *
+     * @param {string} moduleSpecifier - A unique module specifier, used to identify the module.
+     * @param {Function} ModuleClass - The ES Script Module class to add to the component
+     * @param {AttributeDefinition} attributeDefinition - A definition of the attributes used in the Script ESM class
+     * @param {object} [attributes] - A set of attributes which are assigned to the Script Module instance
+     * @returns {ModuleInstance} An instance of the module
+     * @private
+     */
+    create(moduleSpecifier, ModuleClass, attributeDefinition = {}, attributes = {}) {
+
+        if (!ModuleClass)
+            throw new Error(`The module '${moduleSpecifier}' does not export a default`);
+
+        if (typeof ModuleClass !== 'function')
+            throw new Error(`The module '${moduleSpecifier}' does not export a class or a function`);
+
+        const moduleInstance = new ModuleClass(this.system.app, this.entity);
+
+        // Assign any attribute definition that have been provided, or if not, assign the default
+        EsmScriptComponent.populateWithAttributes(moduleInstance, attributeDefinition, attributes);
+
+        // Retrieve any previous instance associated with the module specifier
+        const previousModuleInstance = this.modules.get(moduleSpecifier)?.moduleInstance;
+        const isHMREnabled = moduleInstance.swap && typeof moduleInstance.swap === 'function';
+
+        // Add Modules to relevant update/post-update lists
+        if (doesClassHaveMethod(ModuleClass, 'update')) this.modulesWithUpdate.add(moduleSpecifier);
+        if (doesClassHaveMethod(ModuleClass, 'postUpdate')) this.modulesWithPostUpdate.add(moduleSpecifier);
+
+        this.fire('create', moduleSpecifier, moduleInstance);
+        this.fire('create:' + moduleSpecifier, moduleInstance);
```

Comment:
do we need to fire this twice?

Reply:
looking down, it seems one is for the user of this module itself, so probably yes.